intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/method/KotlinMethods.kt

```diff
@@ -0,0 +1,137 @@
+   *     LOCALVARIABLE anInternalType Linternal/defaultMethod/AnInternalType; L0 L2 1
+   *     LOCALVARIABLE s Ljava/lang/String; L0 L2 2
+   *     MAXSTACK = 3
+   *     MAXLOCALS = 3
+   * ```
+   *
+   * @see [`JvmDefault` annotation](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/jvm/src/kotlin/jvm/JvmDefault.kt)
+   */
+  fun Method.isKotlinDefaultMethod(): Boolean {
+    val method: Method = this
+    return cache.computeIfAbsent(method.location) {
+      return@computeIfAbsent isKotlinMethodInvokingDefaultImpls(method)
+    }
+  }
+
+  private fun Method.isKotlinMethodInvokingDefaultImpls(method: Method): Boolean {
+    // filter non kotlin classes
+    if (!method.containingClassFile.annotations.any { it.desc == "Lkotlin/Metadata;" }) {
+      return false
+    }
+
+    // Sanity check : if the method does not have bytecode
+    // this heuristic cannot run
+    if (instructions.isEmpty()) {
+      return false
+    }
+
+    // skip non opcodes, and kotlin intrinsics on arguments
+    val candidateOpcodes = mutableListOf<AbstractInsnNode>()
+    var i = 0
+    do {
+      val currentInsnNode = instructions[i]
+      if (currentInsnNode.opcode == -1) {
+        continue
+      }
+
+      // Drop kotlin Intrinsics
+      if (currentInsnNode.opcode == Opcodes.ALOAD
+        && currentInsnNode.next?.opcode == Opcodes.LDC
+        && (currentInsnNode.next?.next?.opcode == Opcodes.INVOKESTATIC
+          && (currentInsnNode.next?.next as MethodInsnNode).owner == "kotlin/jvm/internal/Intrinsics")) {
+        i += 2
+        continue
+      }
+
+      candidateOpcodes.add(currentInsnNode)
+    } while (++i < instructions.size)
+
+
+    val expectedOpcodes = (
+      3 // aload this + invokestatic + (return or areturn)
+        + method.methodParameters.size // aload for each parameter
+      )
+
+    if (candidateOpcodes.size != expectedOpcodes
+      || candidateOpcodes[0].opcode != Opcodes.ALOAD // aload this
+      || candidateOpcodes.slice(1..method.methodParameters.size).any() { it.opcode != Opcodes.ALOAD } // parameters
+      || candidateOpcodes[candidateOpcodes.lastIndex - 1].opcode != Opcodes.INVOKESTATIC
+      || candidateOpcodes.last().opcode !in intArrayOf(
+        Opcodes.RETURN,
+        Opcodes.ARETURN,
+        Opcodes.DRETURN,
+        Opcodes.FRETURN,
+        Opcodes.IRETURN,
+        Opcodes.LRETURN
+      )
+    ) {
+      return false
+    }
+
+    val methodInsnNode = candidateOpcodes[candidateOpcodes.lastIndex - 1] as MethodInsnNode
+    if (methodInsnNode.name != method.name || !methodInsnNode.owner.endsWith("\$DefaultImpls")) {
+      return false
+    }
+
+    val actualKotlinOwner = methodInsnNode.owner.substring(0, methodInsnNode.owner.length - "\$DefaultImpls".length)
+
+    // do we want to walk the whole class hierarchy?
+    val isAParent = method.containingClassFile.interfaces.any {
+      it == actualKotlinOwner
+    }
+    if (!isAParent) {
+      return false
```

Comment:
Just want to confirm: is it expected behavior that this check will filter out methods which are generated into the `DefaultImpls` classes of interfaces and delegate to the superinterface?

So for the example above
```
interface A {
    fun foo() {}
}

interface B : A

class C : B
```

For `C.foo`, `isKotlinMethodInvokingDefaultImpls` will return true, which is good.

But there's also `B$DefaultImpls.foo`, for which `isKotlinMethodInvokingDefaultImpls` will return false because `B$DefaultImpls` does not have any superclasses, and this is expected, right?

Reply:
> But there's also `B$DefaultImpls.foo`, for which `isKotlinMethodInvokingDefaultImpls` will return false because `B$DefaultImpls` does not have any superclasses, and this is expected, right?

Ah I didn't thought about this, indeed looking at `B$DefaultImpls`' bytecode, super is `Object` and doesn't have interfaces. 

The `B$DefaultImpls.foo` implementation
```
// access flags 0x9
public static foo(Linternal/defaultMethod/B;)Linternal/defaultMethod/AnInternalType;
  // annotable parameter count: 1 (invisible)
  @Lorg/jetbrains/annotations/NotNull;() // invisible, parameter 0
 L0
  LINENUMBER 3 L0
  ALOAD 0
  CHECKCAST internal/defaultMethod/A
  INVOKESTATIC internal/defaultMethod/A$DefaultImpls.foo (Linternal/defaultMethod/A;)Linternal/defaultMethod/AnInternalType;
 L1
  LINENUMBER 10 L1
  ARETURN
 L2
  LOCALVARIABLE $this Linternal/defaultMethod/B; L0 L2 0
  MAXSTACK = 1
  MAXLOCALS = 1
```

I see two cases:
1. If both interfaces `A` and `B` are part of IDEA, then it won't be an issue, because `B` / `B$DefaultImpls` shouldn't be analyzed by the verifier.
2. However if the plugin has an in-between-interface `I` that extends `B`, then this might be an issue, because `I$DefautImpls` will not be filtered out.

Do you think I should introduce additional checks for this specific case, the class is named `xxx$DefaultImpls` and calls another `...$DefaultImpls` ?
I'm ok with that.

Reply:
@LChernigovskaya Fantastic, I would definitely like to support case 2 as I think it's doable. I just didn't yet started.

Reply:
See #976